// koa-swagger-decorator/src/myApi.ts
/* eslint-disable */
/* tslint:disable */
/*
 * ---------------------------------------------------------------
 * ## THIS FILE WAS GENERATED VIA SWAGGER-TYPESCRIPT-API        ##
 * ##                                                           ##
 * ## AUTHOR: acacode                                           ##
 * ## SOURCE: https://github.com/acacode/swagger-typescript-api ##
 * ---------------------------------------------------------------
 */

export type QueryParamsType = Record<string | number, any>;
export type ResponseFormat = keyof Omit<Body, "body" | "bodyUsed">;

export interface FullRequestParams extends Omit<RequestInit, "body"> {
  /** set parameter to `true` for call `securityWorker` for this request */
  secure?: boolean;
  /** request path */
  path: string;
  /** content type of request body */
  type?: ContentType;
  /** query params */
  query?: QueryParamsType;
  /** format of response (i.e. response.json() -> format: "json") */
  format?: ResponseFormat;
  /** request body */
  body?: unknown;
  /** base url */
  baseUrl?: string;
  /** request cancellation token */
  cancelToken?: CancelToken;
}

export type RequestParams = Omit<FullRequestParams, "body" | "method" | "query" | "path">;

export interface ApiConfig<SecurityDataType = unknown> {
  baseUrl?: string;
  baseApiParams?: Omit<RequestParams, "baseUrl" | "cancelToken" | "signal">;
  securityWorker?: (securityData: SecurityDataType | null) => Promise<RequestParams | void> | RequestParams | void;
  customFetch?: typeof fetch;
}

export interface HttpResponse<D extends unknown, E extends unknown = unknown> extends Response {
  data: D;
  error: E;
}

type CancelToken = Symbol | string | number;

export enum ContentType {
  Json = "application/json",
  FormData = "multipart/form-data",
  UrlEncoded = "application/x-www-form-urlencoded",
}

export class HttpClient<SecurityDataType = unknown> {
  public baseUrl: string = "";
  private securityData: SecurityDataType | null = null;
  private securityWorker?: ApiConfig<SecurityDataType>["securityWorker"];
  private abortControllers = new Map<CancelToken, AbortController>();
  private customFetch = (...fetchParams: Parameters<typeof fetch>) => fetch(...fetchParams);

  private baseApiParams: RequestParams = {
    credentials: "same-origin",
    headers: {},
    redirect: "follow",
    referrerPolicy: "no-referrer",
  };

  constructor(apiConfig: ApiConfig<SecurityDataType> = {}) {
    Object.assign(this, apiConfig);
  }

  public setSecurityData = (data: SecurityDataType | null) => {
    this.securityData = data;
  };

  protected encodeQueryParam(key: string, value: any) {
    const encodedKey = encodeURIComponent(key);
    return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : `${value}`)}`;
  }

  protected addQueryParam(query: QueryParamsType, key: string) {
    return this.encodeQueryParam(key, query[key]);
  }

  protected addArrayQueryParam(query: QueryParamsType, key: string) {
    const value = query[key];
    return value.map((v: any) => this.encodeQueryParam(key, v)).join("&");
  }

  protected toQueryString(rawQuery?: QueryParamsType): string {
    const query = rawQuery || {};
    const keys = Object.keys(query).filter((key) => "undefined" !== typeof query[key]);
    return keys
      .map((key) => (Array.isArray(query[key]) ? this.addArrayQueryParam(query, key) : this.addQueryParam(query, key)))
      .join("&");
  }

  protected addQueryParams(rawQuery?: QueryParamsType): string {
    const queryString = this.toQueryString(rawQuery);
    return queryString ? `?${queryString}` : "";
  }

  private contentFormatters: Record<ContentType, (input: any) => any> = {
    [ContentType.Json]: (input: any) =>
      input !== null && (typeof input === "object" || typeof input === "string") ? JSON.stringify(input) : input,
    [ContentType.FormData]: (input: any) =>
      Object.keys(input || {}).reduce((formData, key) => {
        const property = input[key];
        formData.append(
          key,
          property instanceof Blob
            ? property
            : typeof property === "object" && property !== null
            ? JSON.stringify(property)
            : `${property}`,
        );
        return formData;
      }, new FormData()),
    [ContentType.UrlEncoded]: (input: any) => this.toQueryString(input),
  };

  protected mergeRequestParams(params1: RequestParams, params2?: RequestParams): RequestParams {
    return {
      ...this.baseApiParams,
      ...params1,
      ...(params2 || {}),
      headers: {
        ...(this.baseApiParams.headers || {}),
        ...(params1.headers || {}),
        ...((params2 && params2.headers) || {}),
      },
    };
  }

  protected createAbortSignal = (cancelToken: CancelToken): AbortSignal | undefined => {
    if (this.abortControllers.has(cancelToken)) {
      const abortController = this.abortControllers.get(cancelToken);
      if (abortController) {
        return abortController.signal;
      }
      return void 0;
    }

    const abortController = new AbortController();
    this.abortControllers.set(cancelToken, abortController);
    return abortController.signal;
  };

  public abortRequest = (cancelToken: CancelToken) => {
    const abortController = this.abortControllers.get(cancelToken);

    if (abortController) {
      abortController.abort();
      this.abortControllers.delete(cancelToken);
    }
  };

  public request = async <T = any, E = any>({
    body,
    secure,
    path,
    type,
    query,
    format,
    baseUrl,
    cancelToken,
    ...params
  }: FullRequestParams): Promise<HttpResponse<T, E>> => {
    const secureParams =
      ((typeof secure === "boolean" ? secure : this.baseApiParams.secure) &&
        this.securityWorker &&
        (await this.securityWorker(this.securityData))) ||
      {};
    const requestParams = this.mergeRequestParams(params, secureParams);
    const queryString = query && this.toQueryString(query);
    const payloadFormatter = this.contentFormatters[type || ContentType.Json];
    const responseFormat = format || requestParams.format;

    return this.customFetch(`${baseUrl || this.baseUrl || ""}${path}${queryString ? `?${queryString}` : ""}`, {
      ...requestParams,
      headers: {
        ...(type && type !== ContentType.FormData ? { "Content-Type": type } : {}),
        ...(requestParams.headers || {}),
      },
      signal: cancelToken ? this.createAbortSignal(cancelToken) : requestParams.signal,
      body: typeof body === "undefined" || body === null ? null : payloadFormatter(body),
    }).then(async (response) => {
      const r = response as HttpResponse<T, E>;
      r.data = null as unknown as T;
      r.error = null as unknown as E;

      const data = !responseFormat
        ? r
        : await response[responseFormat]()
            .then((data) => {
              if (r.ok) {
                r.data = data;
              } else {
                r.error = data;
              }
              return r;
            })
            .catch((e) => {
              r.error = e;
              return r;
            });

      if (cancelToken) {
        this.abortControllers.delete(cancelToken);
      }

      if (!response.ok) throw data;
      return data;
    });
  };
}

/**
 * @title API DOC
 * @version 1.0.0
 *
 * API DOC
 */
export class Api<SecurityDataType extends unknown> extends HttpClient<SecurityDataType> {
  user = {
    /**
     * @description example of api
     *
     * @tags User
     * @name RegisterCreate
     * @summary register user
     * @request POST:/user/register
     * @secure
     */
    registerCreate: (data: { name: string; password: string }, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/user/register`,
        method: "POST",
        body: data,
        secure: true,
        ...params,
      }),

    /**
     * @description user login, password is 123456
     *
     * @tags User
     * @name LoginCreate
     * @summary user login, password is 123456
     * @request POST:/user/login
     * @secure
     */
    loginCreate: (data: { name: string; password: string }, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/user/login`,
        method: "POST",
        body: data,
        secure: true,
        ...params,
      }),

    /**
     * @description user list
     *
     * @tags User
     * @name UserList
     * @summary user list
     * @request GET:/user
     * @secure
     */
    userList: (params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/user`,
        method: "GET",
        secure: true,
        ...params,
      }),

    /**
     * @description get user by id
     *
     * @tags User
     * @name UserDetail
     * @summary get user by id
     * @request GET:/user/{id}
     * @secure
     */
    userDetail: (id: string, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/user/${id}`,
        method: "GET",
        secure: true,
        ...params,
      }),

    /**
     * @description delete user by id
     *
     * @tags User
     * @name UserDelete
     * @summary delete user by id
     * @request DELETE:/user/{id}
     * @secure
     */
    userDelete: (id: string, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/user/${id}`,
        method: "DELETE",
        secure: true,
        ...params,
      }),
  };
  sample = {
    /**
     * @description exampling [formdata] and [middlewares] decorators
     *
     * @tags A
     * @name SampleCreate
     * @summary showing upload files example using koa-multer
     * @request POST:/sample
     * @deprecated
     * @secure
     */
    sampleCreate: (query: { page: number; myLimit?: number }, data: { file?: File }, params: RequestParams = {}) =>
      this.request<void, void>({
        path: `/sample`,
        method: "POST",
        query: query,
        body: data,
        secure: true,
        type: ContentType.FormData,
        ...params,
      }),
  };
  enum = {
    /**
     * @description example of  enum
     *
     * @tags B, A
     * @name EnumList
     * @summary example of  enum
     * @request GET:/enum
     * @secure
     */
    enumList: (query: { page?: "1" | "2"; limit: number }, params: RequestParams = {}) =>
      this.request<void, { id?: number; name?: string }>({
        path: `/enum`,
        method: "GET",
        query: query,
        secure: true,
        ...params,
      }),
  };
  v1 = {
    /**
     * @description class decorator for url prefix
     *
     * @tags prefix
     * @name PrefixList
     * @summary class decorator for url prefix
     * @request GET:/v1/prefix
     * @secure
     */
    prefixList: (query?: { page?: string }, params: RequestParams = {}) =>
      this.request<void, void>({
        path: `/v1/prefix`,
        method: "GET",
        query: query,
        secure: true,
        ...params,
      }),
  };
  egg = {
    /**
     * @description egg-like controller method1
     *
     * @tags Egg
     * @name Method1List
     * @summary egg-like controller method1
     * @request GET:/egg/method1
     * @secure
     */
    method1List: (query?: { no?: string }, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/egg/method1`,
        method: "GET",
        query: query,
        secure: true,
        ...params,
      }),

    /**
     * @description egg-like controller method2
     *
     * @tags Egg
     * @name Method2Create
     * @summary egg-like controller method2
     * @request POST:/egg/method2
     * @secure
     */
    method2Create: (data: { yes?: string | null }, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/egg/method2`,
        method: "POST",
        body: data,
        secure: true,
        type: ContentType.Json,
        ...params,
      }),
  };
  other = {
    /**
     * @description something in sub routes
     *
     * @tags Other
     * @name OtherList
     * @summary something in sub routes
     * @request GET:/other
     * @secure
     */
    otherList: (params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/other`,
        method: "GET",
        secure: true,
        ...params,
      }),

    /**
     * @description validator test
     *
     * @tags Other
     * @name OtherCreate
     * @summary validator test
     * @request POST:/other
     * @secure
     */
    otherCreate: (
      data: {
        str?: string;
        boo?: boolean;
        nn?: number;
        foo: undefined[];
        bar: { url?: string; name?: string }[];
        exampleItem: { url?: string; name?: string };
        obb?: { aaaa?: string; bbbb?: string };
      },
      params: RequestParams = {},
    ) =>
      this.request<void, any>({
        path: `/other`,
        method: "POST",
        body: data,
        secure: true,
        type: ContentType.Json,
        ...params,
      }),

    /**
     * @description something else in sub routes
     *
     * @tags Other
     * @name WhatList
     * @summary something else in sub routes
     * @request GET:/other/what
     * @secure
     */
    whatList: (query?: { no?: any }, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/other/what`,
        method: "GET",
        query: query,
        secure: true,
        ...params,
      }),

    /**
     * @description test when type is not defined
     *
     * @tags Other
     * @name WhatCreate
     * @summary test when type is not defined
     * @request POST:/other/what
     * @secure
     */
    whatCreate: (data: { num?: number; ss?: any; xx?: object }, params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/other/what`,
        method: "POST",
        body: data,
        secure: true,
        type: ContentType.Json,
        ...params,
      }),

    /**
     * @description something in sub sub routes
     *
     * @tags PP
     * @name GetOther
     * @summary something in sub sub routes
     * @request GET:/other/how
     * @secure
     */
    getOther: (params: RequestParams = {}) =>
      this.request<void, any>({
        path: `/other/how`,
        method: "GET",
        secure: true,
        ...params,
      }),
  };
}
