// koa-swagger-decorator/dist/wrapper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwaggerRouter = exports.wrapper = void 0;
const router_1 = __importDefault(require("@koa/router"));
const is_type_of_1 = __importDefault(require("is-type-of"));
const validate_1 = __importDefault(require("./validate"));
const swaggerHTML_1 = require("./swaggerHTML");
const swaggerJSON_1 = require("./swaggerJSON");
const swaggerObject_1 = __importDefault(require("./swaggerObject"));
const utils_1 = require("./utils");
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const koa_compose_1 = __importDefault(require("koa-compose"));
const ramda_1 = require("ramda");
const validator = (parameters) => (ctx, next) => __awaiter(void 0, void 0, void 0, function* () {
    if (!parameters) {
        yield next();
        return;
    }
    if (parameters.query) {
        ctx.validatedQuery = (0, validate_1.default)(ctx.request.query, parameters.query);
    }
    if (parameters.path) {
        ctx.validatedParams = (0, validate_1.default)(ctx.params, parameters.path);
    }
    if (parameters.body) {
        ctx.validatedBody = (0, validate_1.default)(ctx.request.body, parameters.body);
    }
    yield next();
});
const handleDumpSwaggerJSON = (router, dumpOptions, options = {}) => {
    let data = {};
    const { dir = process.cwd(), filename = 'swagger.json' } = dumpOptions;
    Object.keys(swaggerObject_1.default.data).forEach((k) => {
        if (router.swaggerKeys.has(k)) {
            data[k] = swaggerObject_1.default.data[k];
        }
    });
    console.log(path_1.default.resolve(dir, filename));
    const jsonData = (0, swaggerJSON_1.swaggerJSON)(options, data);
    (0, fs_1.writeFileSync)(path_1.default.resolve(dir, filename), JSON.stringify(jsonData, null, 2));
};
const handleSwagger = (router, options) => {
    const { swaggerJsonEndpoint = '/swagger-json', swaggerHtmlEndpoint = '/swagger-html', prefix = '', swaggerConfiguration = {}, } = options;
    // setup swagger router
    router.get(swaggerJsonEndpoint, (ctx) => __awaiter(void 0, void 0, void 0, function* () {
        let data = {};
        if (router instanceof SwaggerRouter) {
            Object.keys(swaggerObject_1.default.data).forEach(k => {
                if (router.swaggerKeys.has(k)) {
                    data[k] = swaggerObject_1.default.data[k];
                }
            });
        }
        else {
            // 兼容使用 wrapper 的情况
            data = swaggerObject_1.default.data;
        }
        ctx.body = (0, swaggerJSON_1.swaggerJSON)(options, data);
    }));
    router.get(swaggerHtmlEndpoint, (ctx) => __awaiter(void 0, void 0, void 0, function* () {
        ctx.body = (0, swaggerHTML_1.swaggerHTML)((0, utils_1.getPath)(prefix, swaggerJsonEndpoint), swaggerConfiguration);
    }));
};
const handleMap = (router, SwaggerClass, { doValidation = true }) => {
    if (!SwaggerClass)
        return;
    const classMiddlewares = SwaggerClass.middlewares || [];
    const classPrefix = SwaggerClass.prefix || '';
    const classParameters = SwaggerClass.parameters || {};
    const classParametersFilters = SwaggerClass.parameters
        ? SwaggerClass.parameters.filters
        : ['ALL'];
    classParameters.query = classParameters.query ? classParameters.query : {};
    const staticMethods = Object.getOwnPropertyNames(SwaggerClass)
        .filter(method => !utils_1.reservedMethodNames.includes(method))
        .map((method) => {
        const func = SwaggerClass[method];
        if (typeof func !== 'object' && typeof func !== 'function') {
            return {};
        }
        func['fnName'] = method;
        return func;
    });
    const SwaggerClassPrototype = SwaggerClass.prototype;
    const methods = Object.getOwnPropertyNames(SwaggerClassPrototype)
        .filter((method) => !utils_1.reservedMethodNames.includes(method))
        .map((method) => {
        const wrapperMethod = (ctx) => __awaiter(void 0, void 0, void 0, function* () {
            const c = new SwaggerClass(ctx);
            yield c[method](ctx);
        });
        // 添加了一层 wrapper 之后，需要把原函数的名称暴露出来 fnName
        // wrapperMethod 继承原函数的 descriptors
        const descriptors = Object.getOwnPropertyDescriptors(SwaggerClassPrototype[method]);
        Object.defineProperties(wrapperMethod, Object.assign({ fnName: {
                value: method,
                enumerable: true,
                writable: true,
                configurable: true,
            } }, descriptors));
        return wrapperMethod;
    });
    // map all methods
    [...staticMethods, ...methods]
        // filter methods withour @request decorator
        .filter((item) => {
        const { path, method } = item;
        if (!path && !method) {
            return false;
        }
        return true;
    })
        // add router
        .forEach((item) => {
        router._addKey(`${SwaggerClass.name}-${item.fnName}`);
        const { path, method } = item;
        let { middlewares = [] } = item;
        const localParams = item.parameters || {};
        if (classParametersFilters.includes('ALL') ||
            classParametersFilters.map(i => i.toLowerCase()).includes(method)) {
            const globalQuery = (0, ramda_1.clone)(classParameters.query);
            localParams.query = localParams.query ? localParams.query : {};
            // merge local query and class query
            // local query 的优先级更高
            localParams.query = Object.assign(globalQuery, localParams.query);
        }
        if (is_type_of_1.default.function(middlewares)) {
            middlewares = [middlewares];
        }
        if (!is_type_of_1.default.array(middlewares)) {
            throw new Error('middlewares params must be an array or function');
        }
        middlewares.forEach((item) => {
            if (!is_type_of_1.default.function(item)) {
                throw new Error('item in middlewares must be a function');
            }
        });
        if (!utils_1.allowedMethods.hasOwnProperty(method.toUpperCase())) {
            throw new Error(`illegal API: ${method} ${path} at [${item}]`);
        }
        const chain = [`${(0, utils_1.convertPath)(`${classPrefix}${path}`)}`];
        if (doValidation) {
            chain.push(validator(localParams));
        }
        chain.push(...classMiddlewares);
        chain.push(...middlewares);
        chain.push(item);
        router[method](...chain);
    });
};
const handleMapDir = (router, dir, options) => {
    (0, utils_1.loadSwaggerClasses)(dir, options).forEach((c) => {
        router.map(c, options);
    });
};
const handleBuildMiddleware = (router, SwaggerClass, { doValidation = true }) => {
    const meta = [];
    if (!SwaggerClass)
        return meta;
    const classMiddlewares = SwaggerClass.middlewares || [];
    const classPrefix = SwaggerClass.prefix || '';
    const classParameters = SwaggerClass.parameters || {};
    const classParametersFilters = SwaggerClass.parameters
        ? SwaggerClass.parameters.filters
        : ['ALL'];
    classParameters.query = classParameters.query ? classParameters.query : {};
    const staticMethods = Object.getOwnPropertyNames(SwaggerClass)
        .filter(method => !utils_1.reservedMethodNames.includes(method))
        .map((method) => {
        const func = SwaggerClass[method];
        func['fnName'] = method;
        return func;
    });
    const SwaggerClassPrototype = SwaggerClass.prototype;
    const methods = Object.getOwnPropertyNames(SwaggerClassPrototype)
        .filter((method) => !utils_1.reservedMethodNames.includes(method))
        .map((method) => {
        const wrapperMethod = (ctx) => __awaiter(void 0, void 0, void 0, function* () {
            const c = new SwaggerClass(ctx);
            yield c[method](ctx);
        });
        // 添加了一层 wrapper 之后，需要把原函数的名称暴露出来 fnName
        // wrapperMethod 继承原函数的 descriptors
        const descriptors = Object.getOwnPropertyDescriptors(SwaggerClassPrototype[method]);
        Object.defineProperties(wrapperMethod, Object.assign({ fnName: {
                value: method,
                enumerable: true,
                writable: true,
                configurable: true,
            } }, descriptors));
        return wrapperMethod;
    });
    // map all methods
    [...staticMethods, ...methods]
        // filter methods withour @request decorator
        .filter((item) => {
        const { path, method } = item;
        if (!path && !method) {
            return false;
        }
        return true;
    })
        // add router
        .forEach((item) => {
        router._addKey(`${SwaggerClass.name}-${item.fnName}`);
        const { path, method } = item;
        let { middlewares = [] } = item;
        const localParams = item.parameters || {};
        if (classParametersFilters.includes('ALL') ||
            classParametersFilters.map(i => i.toLowerCase()).includes(method)) {
            const globalQuery = (0, ramda_1.clone)(classParameters.query);
            localParams.query = localParams.query ? localParams.query : {};
            // merge local query and class query
            // local query 的优先级更高
            localParams.query = Object.assign(globalQuery, localParams.query);
        }
        if (is_type_of_1.default.function(middlewares)) {
            middlewares = [middlewares];
        }
        if (!is_type_of_1.default.array(middlewares)) {
            throw new Error('middlewares params must be an array or function');
        }
        middlewares.forEach((item) => {
            if (!is_type_of_1.default.function(item)) {
                throw new Error('item in middlewares must be a function');
            }
        });
        if (!utils_1.allowedMethods.hasOwnProperty(method.toUpperCase())) {
            throw new Error(`illegal API: ${method} ${path} at [${item}]`);
        }
        const endpoint = `${(0, utils_1.convertPath)(`${classPrefix}${path}`)}`;
        const chain = [];
        if (doValidation) {
            chain.push(validator(localParams));
        }
        chain.push(...classMiddlewares);
        chain.push(...middlewares);
        meta.push({ path: endpoint, method: method.toLowerCase(), middlewares: chain, name: item.name });
    });
    return meta.map(m => ({
        path: m.path, method: m.method, middleware: (0, koa_compose_1.default)(m.middlewares), name: m.name
    }));
};
const wrapper = (router) => {
    router.swagger = (options = {}) => {
        handleSwagger(router, options);
    };
    router.map = (SwaggerClass, options = {}) => {
        handleMap(router, SwaggerClass, options);
    };
    router.mapDir = (dir, options = {}) => {
        handleMapDir(router, dir, options);
    };
};
exports.wrapper = wrapper;
class SwaggerRouter extends router_1.default {
    constructor(opts = {}, swaggerOpts = {}) {
        super(opts);
        this.opts = opts || {}; // koa-router opts
        this.swaggerKeys = new Set();
        this.swaggerOpts = swaggerOpts || {}; // swagger-router opts
        if (this.opts.prefix && !this.swaggerOpts.prefix) {
            this.swaggerOpts.prefix = this.opts.prefix;
        }
    }
    _addKey(str) {
        this.swaggerKeys.add(str);
    }
    swagger(options = {}) {
        const opts = Object.assign(options, this.swaggerOpts);
        handleSwagger(this, opts);
    }
    dumpSwaggerJson(dumpOptions, options = {}) {
        const opts = Object.assign(options, this.swaggerOpts);
        handleDumpSwaggerJSON(this, dumpOptions, opts);
    }
    map(SwaggerClass, options) {
        handleMap(this, SwaggerClass, options);
    }
    mapDir(dir, options = {}) {
        handleMapDir(this, dir, options);
    }
    // compose & create a middleware for validator & @middlewares decorators
    buildMiddleware(SwaggerClass, options) {
        return handleBuildMiddleware(this, SwaggerClass, options);
    }
}
exports.SwaggerRouter = SwaggerRouter;
